pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -0,0 +1,435 @@
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        type Currency: Inspect<Self::AccountId> + Balanced<Self::AccountId>;
+
+        type XcmSender: SendXcm;
+        type GetPurchaseCoreCall: GetPurchaseCoreCall;
+        type GetBlockNumber: Get<u32>;
+        type GetParathreadAccountId: Convert<ParaId, [u8; 32]>;
+        type SelfParaId: Get<ParaId>;
+        type MaxParathreads: Get<u32>;
+        type GetParathreadParams: GetParathreadParams;
+        type GetAssignedCollators: GetParathreadCollators<Self::AccountId>;
+        /// A configuration for base priority of unsigned transactions.
+        ///
+        /// This is exposed so that it can be tuned for particular runtime, when
+        /// multiple pallets send unsigned transactions.
+        #[pallet::constant]
+        type UnsignedPriority: Get<TransactionPriority>;
+
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// An XCM message to buy a core for this parathread has been sent to the relay chain.
+        BuyCoreXcmSent { para_id: ParaId },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        InvalidProof,
+        ErrorValidatingXCM,
+        ErrorDeliveringXCM,
+        /// An order for this para id already exists
+        OrderAlreadyExists,
+        /// The para id is not a parathread
+        NotAParathread,
+        /// There are too many in-flight orders, buying cores will not work until some of those
+        /// orders finish.
+        InFlightLimitReached,
+        /// There are no collators assigned to this parathread, so no point in buying a core
+        NoAssignedCollators,
+        /// This collator is not assigned to this parathread
+        CollatorNotAssigned,
+        /// The `XcmWeights` storage has not been set. This must have been set by root with the
+        /// value of the relay chain xcm call weight and extrinsic weight
+        XcmWeightStorageNotSet,
+    }
+
+    /// Proof that I am a collator, assigned to a para_id, and I can buy a core for that para_id
+    #[derive(Encode, Decode, CloneNoBound, PartialEq, Eq, DebugNoBound, TypeInfo)]
+    #[scale_info(skip_type_params(T))]
+    pub struct BuyCoreCollatorProof<T: Config> {
+        account: T::AccountId,
+        // TODO
+        _signature: (),
+    }
+
+    /// Set of parathreads that have already sent an XCM message to buy a core recently.
+    /// Used to avoid 2 collators buying a core at the same time, because it is only possible to buy
+    /// 1 core in 1 relay block.
+    #[pallet::storage]
+    //pub type InFlightOrders<T: Config> = StorageMap<_, Blake2_128Concat, ParaId, (), OptionQuery>;
```

Comment:
Are we using the `StorageValue` instead of the commented `StorageMap` at the end? If so, we can remove this line.
```suggestion
```